### ISSUE TITLE
Assertion `plan.second.size() <= nr_ports' failed.

### ISSUE THREAD

**Issue by frankystone** [widelands/widelands#5593]
During a training game where the AI build a port i get this assertion:

~~~~
widelands: /home/kaputtnik/Quellcode/widelands-repo/widelands/src/economy/shipping_schedule.cc:779: Duration Widelands::ShippingSchedule::update(Widelands::Game &): Assertion `plan.second.size() <= nr_ports' failed.
~~~~

**To Reproduce**
1. Open attached save game 
2. Attack the barbarian port

Close after the port burned down the assertion happens.

**Version:**
 - OS: archlinux
 - Widelands Version: 1.1~git25864 (b301eaa@HEAD) Debug

**Additional context**
One time this happens also by attacking and conquering a military site which was not close to the port. Don't know, maybe the port get dismantled by the AI?

[test.zip](https://github.com/widelands/widelands/files/9709506/test.zip)

**Comment by Noordfrees** [comment]
I cannot reproduce. If you can trigger it again, can you run Widelands with `--verbose` and attach the full output (this will be a *lot* of text)?

**Comment by frankystone** [comment]
Did you wait long enough? On first try with verbose the assertion does not happen, maybe i also didn't waited enough. On second try i just played further until the assertion get triggered and this is the resulting verbose output:

[ships_assertion.txt](https://github.com/widelands/widelands/files/9718243/ships_assertion.txt)

**Comment by Noordfrees** [comment]
Okay, this is a weird bug. The exact time when the bug hits is different on every run. It only happens when destroying the port, not when leaving this port intact – *however*, this port is in a different fleet from all other ports and ships and therefore should not have any connection to them whatsoever. Indeed, after attacking the port its fleet and schedule are destroyed and there's no rerouting of any sort in the other fleet.

The schedule at the time of the assert looks like this:
```
[0] = {dock = {m = {serial_ = 980}}, expedition = false, load_there = std::map with 1 element = {[{m = {serial_ = 712}}] = 3}, duration_from_previous_location = {
      value_ = std::atomic<unsigned int> = { 2400 }}}, 
[1] = {dock = {m = {serial_ = 3633}}, expedition = false, load_there = std::map with 1 element = {[{m = {serial_ = 980}}] = 1}, 
    duration_from_previous_location = {value_ = std::atomic<unsigned int> = { 14400 }}}, 
[2] = {dock = {m = {serial_ = 980}}, expedition = false, load_there = std::map with 0 elements, 
    duration_from_previous_location = {value_ = std::atomic<unsigned int> = { 14400 }}}, 
[3] = {dock = {m = {serial_ = 712}}, expedition = false, load_there = std::map with 0 elements, 
    duration_from_previous_location = {value_ = std::atomic<unsigned int> = { 1800 }}}}
```
This is the schedule for the *Dellingr*, the only ship in a fleet at the time. Indeed there is a duplicate port there (ID 980 at coords 80×23), which should never ever happen. 

However it may actually be valid here – the plan is that we first go to 980, load cargo for 712, then go to 3633, load cargo for 980, then go to 980 *again* to deliver it, and then to 712 to deliver the other cargo. There are two ways to deal with this: either relax the assumptions and permit such plans, or ensure we can't load cargo for a destination that comes earlier in the current plan than the starting port.
The second way would be less efficient in small fleets but more robust in large ones, so that's probably the way to go.

**Comment by frankystone** [comment]
> Okay, this is a weird bug. The exact time when the bug hits is different on every run. It only happens when destroying the port, not when leaving this port intact

No, you can just do nothing or conquer some of the nearby barbarian sentries and after some time the assert happens (Ai didn't destroys/dismatles the port).

Is this port ever reachable by ships (if the shipyard is outside this part of the map)? It looks to me that no ships can enter this part of the water because the space is to narrow near the positions 85/56 and 86/57. So this port should may not be considered in any shipping algorithm?

**Comment by Noordfrees** [comment]
The red player has two separate fleets in separate oceans here: One with just this one port (the one you attack), and one with several ports and a ship. The second, big fleet is the one with a shipyard. The port-to-attack is in a separate fleet, unreachable by ship from the other ports, and therefore has no influence whatsoever on the big fleet's scheduling. 
So the attack is actually completely unrelated to this bug, which is simply triggered by the right combination of economic transfers at the right time (which in turn depends on the AI's decisions which may be different on every run).

**Comment by frankystone** [comment]
Thanks for clarification :)
